[WIP] Parametric Logarithmic Derivative: using Risch structure Theorem

gxyd opened the pull request:
<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

gxyd committed:
Parametric Logarithmic Derivative, using Risch sctructure theorem

gxyd committed:
improve basis_Q

gxyd committed:
update basis_Q function

gxyd committed:
Update function basis_Q

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1042,28 @@ def is_deriv_k(fa, fd, DE):
+    T = set([num**(i.as_coeff_exponent(num)[1])
+        for num in trans_nums for i in expr_args])
+    cons = T.union(set(K.gens))
+    return sfield(list(cons), extension=True)
```
It raises error when the flag `extension=True` is there.

gxyd committed:
Improve the base_Q function and parametric_log_deriv

gxyd committed:
improve base_Q

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1049,47 @@ def is_deriv_k(fa, fd, DE):
+    alg_ind_trans_consts = set()
+
+    for poly in G:
+        trans_consts = list(set(poly.gens) - set([poly.gen]))
```
What should be assumed of `poly`. How many generators and which kind of `domain` should it have?

gxyd replied:
Atleast domain shouldn't be `EX`. It could be an instance of `AlgebraicField` or `PythonRationalField`.

As for the generators, I think it could be `pi` , `E` i.e. the transcendental numbers should need to be there (but the algebraic number not, as `extension=True` should be there with each poly).

jksuom replied:
I wonder why all generators except for the principal one (`poly.gen`) are taken. They could probably all be considered transcendental for computing the rational relations. (It probably does not matter whether they are constants or not in the differential extension. This question is purely algebraic.)

gxyd replied:
> I wonder why all generators except for the principal one (poly.gen) are taken.

Here "all generators" -> "transcendental generators" ?
As for not including `poly.gen`, I think you are right, I should include it. But I think I haven't found a good enough test case for this function. I'll try to do that first and see if it works okay.

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -918,14 +918,25 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
+    C, K, trans_consts = base_Q(DE.D)
+    # trans_consts be t1, ..., tn
+    E_part = [frac_in(DE.D[i].quo(Poly(DE.T[i], DE.T[i])),
+        trans_consts) for i in DE.indices('exp')]
```
What is the result of `frac_in` when `trans_consts` is a list?

gxyd replied:
An example would be
```python
In [4]: frac_in(x**2/(x + 1), [x, t, sqrt(2)])
Out[4]: (Poly(x**2, x, t, sqrt(2), domain='ZZ'), Poly(x + 1, x, t, sqrt(2), domain='ZZ'))
```
So it uses all the elements of list as generator for `num` and `den`.

gxyd committed:
Add a doctest and don't remove poly.gen from trans_consts

gxyd commented:
I have updated the PR. I have also added one doctest example, showing how things work for now with the function `parametric_log_deriv`. (I wouldn't update the PR if it bothers you in the PR reviewing, while you are doing it).

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1062,47 @@ def is_deriv_k(fa, fd, DE):
+    alg_ind_trans_consts = set()
+
+    for poly in G:
+        trans_consts = list(set(poly.gens))
```
Could this be `poly.gens`?

gxyd replied:
Yes, I think that would be okay.

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1062,47 @@ def is_deriv_k(fa, fd, DE):
+        # for ex. if π and π^2 both are in poly.as_expr() then
+        # both should be added
+        T = set([num**(i.as_coeff_exponent(num)[1])
+            for num in trans_consts for i in expr_args])
```
These seem to be the monomials as expressions. Would it be possible to use `monoms`?

jksuom replied:
I did mean the `.monoms()` method of polynomials that does not require transforming to expressions.

gxyd committed:
Fix imports and fix according to comments

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -214,24 +215,24 @@ def poly_linear_constraints(p, d):
+    or a logarithmic derivative of a K-radical using the structure
+    theorem approach.
 
-    Because Poly does not play well with Matrix yet, this algorithm assumes that
```
Why did you remove this? Is it not true any more?

gxyd replied:
We have `PolyMatrix` now, so I think we can (I haven't done it) modify the code of `constant_system` to receive input as tuple of Poly's `(poly_num, poly_den)`.

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
 
+def base_Q(G):
+    """
+    >>> G = [Poly(x**2*sqrt(2) + x, x, extension=True),
```
Please add function description to the docstring.

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
             return (ans, result, const)
 
 
+def base_Q(G):
```
Should be `basis_Q`, not `base_Q` (if this does what it looks like it does).

asmeurer replied:
Can you just use the "parallel" constructor in the polys to do this work for you?

asmeurer commented on sympy/integrals/prde.py:
```diff
@@ -1038,6 +1063,45 @@ def is_deriv_k(fa, fd, DE):
+    for poly in G:
+        monoms = poly.as_expr().args
+        # T = [t1, t2, ..., tn]
+        # obtaining algebraically independent
```
What does "obtaining" mean here? 

And I don't understand this comment because π and π^2 are not algebraically independent.

gxyd replied:
That is wrong what I did here, basically it was earlier calculating 'linearly' independent transcendental constants. I have improved the function to calculate 'algebraically' independent constants.

gxyd committed:
Improve docstring and make changes in 'basis_Q' function

gxyd committed:
use 'sfield' to get transcendental and algebraic generators

gxyd committed:
Merge branch 'master' into param-log-deriv

gxyd committed:
Initial phase of parametric_log_deriv

gxyd commented on sympy/integrals/prde.py:
```diff
@@ -918,15 +919,46 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
+    # each monomial corresponds to a row in the matrix
+    # of entries in 'K' giving the relations.
+    # The elements of F correspond to the columns
+    t = [[num._get_coeff(m) for num in nums] for m in monoms]
```
I'll make this a `PolyMatrix` (imported as `Matrix`).

jksuom replied:
The matrix will be needed to find the nullspace after the entries are made rational.

gxyd committed:
Address comment https://gitter.im/jksuom/sympy?at=5969a1e6329651f46e8bb53b

gxyd committed:
Use Matrix around 'Kmat'

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -918,15 +919,61 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
+            return [zero]*(n - len(rep)) + rep
+
+        # replace each row of Kmat with n rows of Qmat
+        Qmat = Matrix([zip(*[pad(a.rep) for a in row]) for row in Kmat])
```
What is the type of `Qmat[0. 0]`?

gxyd replied:
Type of `Qmat[0, 0]` is `list` (it shouldn't be).

gxyd replied:
Better said to be a `iterator` (which iterates over a list).

jksuom replied:
That is the reason why I suggested a loop `for row in Kmat` where each time n new rows are generated the list is extended, not appended.

gxyd replied:
Yes, I figure that there must be something with that. I am guessing they are separately to be taken as `Matrix` (but please keep the details in hide for a moment).

gxyd committed:
Fix the matrix formation and remove un-necessary tests

gxyd committed:
use 'field=True' instead of 'extension=True'

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -944,7 +945,8 @@ def parametric_log_deriv(fa, fd, wa, wd, DE):
     # find a field containing all the fractions
-    FF, gens = sfield(F, extension=True)
+    # 'field=True' covers the non-algebraic case
+    FF, gens = sfield(F, field=True)
```
`extension=True` should also be there.

gxyd committed:
separate logic of converting to rational linear relations

Create a new function 'linear_relations_in_Q' for its logic

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -917,16 +918,96 @@ def parametric_log_deriv_heu(fa, fd, wa, wd, DE, c1=None):
+    FF, gens = sfield(F, field=True, extension=True)
+    K = FF.domain
+
+    denom = reduce(lambda i, j: i.lcm(j), [gen.denom for gen in gens])
```
`reduce` is no longer a built-in function in Python 3. It has to be imported from functools. (Its use is probably being discouraged.)

gxyd committed:
Complete writing the function 'linear_relation_in_Q'

jksuom commented on sympy/integrals/prde.py:
```diff
@@ -979,6 +992,9 @@ def pad(rep):
         # 'Kmat' is the same of 'Qmat'
         Qmat = Matrix(Kmat)
 
+    V = Qmat.rowspace()
```
The set of linear relations is given by the nullspace of Qmat.

gxyd replied:
Yes, that is true. But wouldn't we be adding after appending `cancel(fa/fd)`?

gxyd replied:
Okay, so should it be like `F` itself is given changed already as input (i.e `fa/fd` is already appending to `F` before input to `linear_relations_in_Q`) and then after that from the all the nullspace matrices we obtain the one put the coresponding `ri` of `fa/fd` to `1`?

jksuom replied:
The idea is to call this function for finding all rational linear relations and then choose those that satisfy some extra conditions such as one coefficient being  -1 as in the case you refer to, and possibly another being an integer. But the caller has to do that processing. `.nullspace` returns a (possibly empty) list of column matrices that are thought of as vectors. They form a basis of the nullspace. The final solutions are suitable linear combinations of the basis vectors.

(The relations can also be obtained from the rowspace, but then extra processing would be needed.)

gxyd committed:
return null space instead of row space

gxyd committed:
extend 'F' with details of unfilled DifferentialExtension

gxyd committed:
convert 'a' to a `Poly` object

An example where this occured is:

>>> e = ((2581284541*exp(t) + 1757211400)*exp(1/(exp(t) + 1) - 10*t)/(39916800*exp(3*t)
            + 119750400*(exp(2*t) + exp(t)) + 39916800))
>>> risch_integrate(e, t)
AttributeError: 'Zero' object has no attribute 'degree'

gxyd committed:
Add test case for 'parametric_log_deriv'

gxyd committed:
import 'parametric_log_deriv' in test file

gxyd committed:
Modify tests

gxyd commented:
There are failures currently:
```python

[gxyd@stallman sympy{param-log-deriv}]$ python3 ./bin/test sympy/integrals/tests/test_rde.py 
============================================================= test process starts =============================================================
executable:         /home/gxyd/anaconda3/bin/python3  (3.6.0-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        67173900
hash randomization: on (PYTHONHASHSEED=4047578247)

sympy/integrals/tests/test_rde.py[10] ...Ef.....                                                                                         [FAIL]

_______________________________________________________________________________________________________________________________________________
____________________________________________ sympy/integrals/tests/test_rde.py:test_special_denom _____________________________________________
  File "/home/gxyd/foss/sympy/sympy/integrals/tests/test_rde.py", line 82, in test_special_denom
    Poly(1, t0), DE) == \
  File "/home/gxyd/foss/sympy/sympy/integrals/rde.py", line 248, in special_denom
    B = ba*pN.quo(bd) + Poly(n, DE.t)*a*derivation(p, DE).quo(p)*pN
  File "/home/gxyd/foss/sympy/sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 4003, in __mul__
    return f.mul(g)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 1391, in mul
    _, per, F, G = f._unify(g)
  File "/home/gxyd/foss/sympy/sympy/polys/polytools.py", line 399, in _unify
    dom, lev = f.rep.dom.unify(g.rep.dom, gens), len(gens) - 1
  File "/home/gxyd/foss/sympy/sympy/polys/domains/domain.py", line 262, in unify
    return K0.unify_with_symbols(K1, symbols)
  File "/home/gxyd/foss/sympy/sympy/polys/domains/domain.py", line 240, in unify_with_symbols
    raise UnificationFailed("can't unify %s with %s, given %s generators" % (K0, K1, tuple(symbols)))
sympy.polys.polyerrors.UnificationFailed: can't unify ZZ with ZZ[x], given (t0, x) generators

================================= tests finished: 8 passed, 1 expected to fail, 1 exceptions, in 2.15 seconds =================================
DO *NOT* COMMIT!
```

gxyd committed:
Fix use of 'a' as a Poly of DE.t at higher level

gxyd commented:
Fixed the above mentioned issue.

gxyd commented on sympy/integrals/tests/test_prde.py:
```diff
@@ -52,7 +53,8 @@ def test_prde_special_denom():
         (Poly(5*x*t + 1, t), Poly(0, t), [(Poly(t, t), Poly(t**2, t)),
         (Poly(2*t, t), Poly(t, t))], Poly(1, x))
-    DE = DifferentialExtension(extension={'D': [Poly(1, x), Poly((t**2 + 1)*2*x, t)]})
+    DE = DifferentialExtension(extension={'D': [Poly(1, x), Poly((t**2 + 1)*2*x, t)]
```
I am guessing we should have test cases for `parametric_log_deriv` corresponding to this (i.e whatever input `parametric_log_deriv` received for this test case)? Since this problem will eventually be tried (and in most of the cases be solved successfully) by "heuristic version" and never reach the "deterministic version" (after we put `try, except` block.

czgdp1807 commented:
@gxyd Are you still interested to work on it? In addition, I want to know which issue this PR was solving?

czgdp1807 commented:
Closing this for now. Please let us know if there is a need to re-open this. Thanks.

oscarbenjamin commented:
@czgdp1807 is there an issue somewhere that tracks progress in Risch? We should reference these closed PRs there rather than just closing them. These represent substantial work that will be lost otherwise.
